pallets/gear-builtin/src/bls12_381.rs

```diff
@@ -0,0 +1,378 @@
+
+pub struct Actor<T: Config>(PhantomData<T>);
+
+impl<T: Config> BuiltinActor for Actor<T> {
+    const ID: u64 = 1;
+
+    type Error = BuiltinActorError;
+
+    fn handle(dispatch: &StoredDispatch, gas_limit: u64) -> (Result<Payload, Self::Error>, u64) {
+        let message = dispatch.message();
+        let payload = message.payload_bytes();
+        let (result, gas_spent) = match payload.first().copied() {
+            Some(REQUEST_MULTI_MILLER_LOOP) => multi_miller_loop::<T>(&payload[1..], gas_limit),
+            Some(REQUEST_FINAL_EXPONENTIATION) => {
+                final_exponentiation::<T>(&payload[1..], gas_limit)
+            }
+            Some(REQUEST_MULTI_SCALAR_MULTIPLICATION_G1) => msm_g1::<T>(&payload[1..], gas_limit),
+            Some(REQUEST_MULTI_SCALAR_MULTIPLICATION_G2) => msm_g2::<T>(&payload[1..], gas_limit),
+            Some(REQUEST_PROJECTIVE_MULTIPLICATION_G1) => {
+                projective_multiplication_g1::<T>(&payload[1..], gas_limit)
+            }
+            Some(REQUEST_PROJECTIVE_MULTIPLICATION_G2) => {
+                projective_multiplication_g2::<T>(&payload[1..], gas_limit)
+            }
+            _ => (Err(BuiltinActorError::DecodingError), 0),
+        };
+
+        (
+            result.map(|response| {
+                response
+                    .encode()
+                    .try_into()
+                    .unwrap_or_else(|_| unreachable!("Response message is too large"))
+            }),
+            gas_spent,
+        )
+    }
+}
+
+fn decode_vec<T: Config, I: Input>(
+    gas_limit: u64,
+    mut gas_spent: u64,
+    input: &mut I,
+) -> (u64, Option<Result<Vec<u8>, BuiltinActorError>>) {
+    let Ok(len) = Compact::<u32>::decode(input).map(u32::from) else {
+        log::debug!(
+            target: LOG_TARGET,
+            "Failed to scale-decode vector length"
+        );
+        return (gas_spent, Some(Err(BuiltinActorError::DecodingError)));
+    };
+
+    let to_spend = <T as Config>::WeightInfo::decode_bytes(len).ref_time();
+    if gas_limit < gas_spent + to_spend {
+        return (gas_spent, None);
+    }
+
+    gas_spent += to_spend;
+
+    let mut items = vec![0u8; len as usize];
+    let bytes_slice = items.as_mut_slice();
+    let result = input.read(bytes_slice).map(|_| items).map_err(|_| {
+        log::debug!(
+            target: LOG_TARGET,
+            "Failed to scale-decode vector data",
+        );
+
+        BuiltinActorError::DecodingError
+    });
+
+    (gas_spent, Some(result))
+}
+
+fn multi_miller_loop<T: Config>(
+    mut payload: &[u8],
+    gas_limit: u64,
+) -> (Result<Response, BuiltinActorError>, u64) {
+    let (gas_spent, result) = decode_vec::<T, _>(gas_limit, 0, &mut payload);
+    let a = match result {
+        Some(Ok(array)) => array,
+        Some(Err(e)) => return (Err(e), gas_spent),
+        None => return (Err(BuiltinActorError::InsufficientGas), gas_spent),
+    };
+
```

Comment:
I guess such things could be simplified. To keep going open an issue of "beautification" of this module pls. Atm ok

Reply:
and about errors: I'd still create some enum here that `: Into<LimitedStr>` to operate with types inside fns only

Reply:
and caller to callee refactor flow

Reply:
#3841